lib/scholar/optimize/golden_section.ex

```diff
@@ -0,0 +1,223 @@
+  """
+
+  import Nx.Defn
+
+  @derive {Nx.Container, containers: [:x, :fun, :converged, :iterations, :fun_evals]}
+  defstruct [:x, :fun, :converged, :iterations, :fun_evals]
+
+  @type t :: %__MODULE__{
+          x: Nx.Tensor.t(),
+          fun: Nx.Tensor.t(),
+          converged: Nx.Tensor.t(),
+          iterations: Nx.Tensor.t(),
+          fun_evals: Nx.Tensor.t()
+        }
+
+  # Golden ratio conjugate: (sqrt(5) - 1) / 2
+  @phi 0.6180339887498949
+
+  opts = [
+    bracket: [
+      type: {:custom, __MODULE__, :__bracket__, []},
+      required: true,
+      doc: """
+      Bracket containing the minimum. A tuple `{a, b}` defining the search interval.
+      The minimum must lie within this interval.
+      """
+    ],
+    tol: [
+      type: {:custom, Scholar.Options, :positive_number, []},
+      default: 1.0e-8,
+      doc: "Absolute tolerance for convergence."
+    ],
+    maxiter: [
+      type: :pos_integer,
+      default: 500,
+      doc: "Maximum number of iterations."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc false
+  def __bracket__(value) do
+    case value do
+      {a, b} when is_number(a) and is_number(b) and a < b ->
+        {:ok, {a, b}}
+
+      {a, b} when is_number(a) and is_number(b) ->
+        {:error,
+         "expected :bracket to be a tuple {a, b} where a < b, got: #{inspect(value)}"}
+
+      _ ->
+        {:error,
+         "expected :bracket to be a tuple {a, b} of numbers, got: #{inspect(value)}"}
+    end
+  end
+
+  @doc """
+  Minimizes a scalar function using golden section search.
+
+  ## Arguments
+
+  * `fun` - The objective function to minimize. Must be a defn-compatible function
+    that takes a scalar tensor and returns a scalar tensor.
+  * `opts` - Options (see below).
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Returns
+
+  A `Scholar.Optimize.GoldenSection` struct with the optimization result:
+
+  * `:x` - The optimal point found
+  * `:fun` - The function value at the optimal point
+  * `:converged` - Whether the optimization converged (1 if true, 0 if false)
+  * `:iterations` - Number of iterations performed
+  * `:fun_evals` - Number of function evaluations
+
+  ## Examples
+
+      iex> fun = fn x -> Nx.pow(Nx.subtract(x, 3), 2) end
+      iex> result = Scholar.Optimize.GoldenSection.minimize(fun, bracket: {0.0, 5.0})
+      iex> Nx.to_number(result.converged)
+      1
+      iex> Nx.all_close(result.x, Nx.tensor(3.0), atol: 1.0e-4) |> Nx.to_number()
+      1
+  """
+  defn minimize(fun, opts \\ []) do
+    {a, b, tol, maxiter} = transform_opts(opts)
+    minimize_n(fun, a, b, tol, maxiter)
+  end
+
+  deftransformp transform_opts(opts) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    {a, b} = opts[:bracket]
+
+    a = Nx.tensor(a, type: :f64)
+    b = Nx.tensor(b, type: :f64)
+    tol = Nx.tensor(opts[:tol], type: :f64)
+    maxiter = Nx.tensor(opts[:maxiter], type: :s64)
+
```

Comment:
I think we can remove these conversions now. Any number will be converted to a tensor when you cross to `defn`, plus if people want to low with lower precision, they should be able to!

```suggestion

    a = Nx.tensor(a, type: :f64)
    b = Nx.tensor(b, type: :f64)
    tol = Nx.tensor(opts[:tol], type: :f64)
    maxiter = Nx.tensor(opts[:maxiter], type: :s64)

```